core/templates/dev/head/services/UserService.js

```diff
@@ -17,48 +17,55 @@
  */
 
 oppia.factory('UserService', [
-  '$http', '$q', '$window', 'UrlInterpolationService', 'UserInfoObjectFactory',
-  'DEFAULT_PROFILE_IMAGE_PATH',
-  function($http, $q, $window, UrlInterpolationService, UserInfoObjectFactory,
-      DEFAULT_PROFILE_IMAGE_PATH) {
+  '$http', '$q', '$window', 'UrlInterpolationService',
+  'UserInfoObjectFactory', 'DEFAULT_PROFILE_IMAGE_PATH',
+  function($http, $q, $window, UrlInterpolationService,
+      UserInfoObjectFactory, DEFAULT_PROFILE_IMAGE_PATH) {
     var PREFERENCES_DATA_URL = '/preferenceshandler/data';
 
     var userInfo = null;
 
     var getUserInfoAsync = function() {
-      if (GLOBALS.userIsLoggedIn) {
-        if (userInfo) {
-          return $q.resolve(userInfo);
-        }
-        return $http.get(
-          '/userinfohandler'
-        ).then(function(response) {
-          userInfo = UserInfoObjectFactory.createFromBackendDict(response.data);
-          return userInfo;
-        });
-      } else {
+      var currentUrl = $window.location.href;
+      if (currentUrl.includes('signup')) {
         return $q.resolve(UserInfoObjectFactory.createDefault());
       }
+      if (userInfo) {
+        return $q.resolve(userInfo);
+      }
+      return $http.get(
+        '/userinfohandler'
+      ).then(function(response) {
+        if (response.data.user_is_logged_in) {
+          userInfo =
+            UserInfoObjectFactory.createFromBackendDict(response.data);
```

Comment:
```suggestion
          userInfo = (
            UserInfoObjectFactory.createFromBackendDict(response.data));
```

Reply:
I think, `/userinfohandler` is called a second time because I got `200 OK` Response after calling `/userinfohandler`. It should be `304 Not Modified` right?
Below is a screenshot.
![cache](https://user-images.githubusercontent.com/12140464/56238721-1d38d680-605d-11e9-90a9-60cd345d78e4.png)

Reply:
Right now it should be called just once I think. Because all the other calls to the service should use the cached (saved in variable for current page) data (if you don't refresh or switch the page).

Reply:
```python
def test_user_info_handler(self):
        """Test the language is saved in the preferences when handler is
        called.
        """
        self.signup(self.EDITOR_EMAIL, self.EDITOR_USERNAME)
        self.login(self.EDITOR_EMAIL)
        json_response = self.get_json('/userinfohandler')
        self.assertDictContainsSubset({
            'is_moderator': False,
            'is_admin': False,
            'is_topic_manager': False,
            'is_super_admin': False,
            'can_create_collections': False,
            'username': self.EDITOR_USERNAME,
            'user_is_logged_in': True}, json_response)
        self.get_json('/userinfohandler',
            expected_status_int=304)
        self.logout()
```
When I used `self.get_json('/userinfohandler', expected_status_int=304)` for second call, it failed because the second call to `/userinfohandler` gives status code 200. So is that means `/userinfohandler` is not cached?

Error : 
``` bash
ERROR: test_user_info_handler (core.controllers.profile_test.UserInfoHandlerTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vmehta342/Documents/OpenSource/oppia/core/controllers/profile_test.py", line 738, in test_user_info_handler
    expected_status_int=304)
  File "/home/vmehta342/Documents/OpenSource/oppia/core/tests/test_utils.py", line 701, in get_json
    status=expected_status_int)
  File "/home/vmehta342/Documents/OpenSource/oppia/../oppia_tools/webtest-1.4.2/webtest/app.py", line 758, in get
    expect_errors=expect_errors)
  File "/home/vmehta342/Documents/OpenSource/oppia/../oppia_tools/webtest-1.4.2/webtest/app.py", line 1120, in do_request
    self._check_status(status, res)
  File "/home/vmehta342/Documents/OpenSource/oppia/../oppia_tools/webtest-1.4.2/webtest/app.py", line 1159, in _check_status
    "Bad response: %s (not %s)", res_status, status)
AppError: Bad response: 200 OK (not 304)

```

Reply:
The response is "cached" in the `userInfo` in this file so you should test it in frontend tests.

Reply:
Okay! @vojtechjelinek can you tell me or give me some resource that help me to write the frontend test for cache thing?

Reply:
Just add test to *UserServiceSpecs.js* that will:
1. Prepare the response data for `/userinfohandler`
1. Call `getUserInfoAsync`
1. Check that the `/userinfohandler` was requested
1. (maybe) Check that the `userInfo` contains data
1. Call `getUserInfoAsync`
1. Check that the `/userinfohandler` WAS NOT requested

Reply:
You can look at how the `should return image data` works

Reply:
Okay!

Reply:
@vojtechjelinek I don't know why python lint issue failed. I haven't changed any python file during this commit

Reply:
@oppia/dev-workflow-team Can you assist, please?

Reply:
@vrajesh978, The lint issue has been solved, so updating your feature branch with upstream/develop will fix the issue.

Reply:
@DubeySandeep Thanks!

Reply:
@DubeySandeep I got the following error while merging the branch to upstream develop
![error](https://user-images.githubusercontent.com/12140464/56868948-6119cd00-69c7-11e9-94b1-fea912d916fb.png)

Reply:
@vrajesh978, recently Apurv has added a message on oppia-dev@ group about this issue and how to fix this. Can you please check whether that helps?

Reply:
Last email in that thread:
![image](https://user-images.githubusercontent.com/16653571/56874518-f95f9280-6a57-11e9-9f66-2a98a59ddd0e.png)